src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2676 @@
+ this.
+-}
+decodeText :: Ptr Word8 -> Word32 -> IO Text
+decodeText ptr len = Text.decodeUtf8 <$> BS.packCStringLen (castPtr ptr, fromIntegral len)
+
+{- |NB: Assumes the data is valid utf8. Protobuf guarantees this, and Rust/tonic
+ does actually implement the validation, so this is safe.
+-}
+decodeReceiveName :: Ptr Word8 -> Word32 -> IO Wasm.ReceiveName
+decodeReceiveName ptr len = Wasm.ReceiveName <$> decodeText ptr len
+
+-- |The result type of a gRPC2 query.
+data QueryResult
+    = -- | An invalid argument was provided by the client.
+      QRInvalidArgument
+    | -- | An internal error occured.
+      QRInternalError
+    | -- | The query succeeded.
+      QRSuccess
+    | -- | The requested data could not be found.
+      QRNotFound
+
+-- |Convert a QueryResult to a result code.
+queryResultCode :: QueryResult -> Int64
+queryResultCode QRInvalidArgument = -2
+queryResultCode QRInternalError = -1
+queryResultCode QRSuccess = 0
+queryResultCode QRNotFound = 1
+
+-- |If the message is given encode and write it using the provided callback.
+returnMessage ::
+    (Proto.Message (Output a), ToProto a) =>
+    (Ptr Word8 -> Int64 -> IO ()) ->
+    -- | The potential message.
+    Maybe a ->
+    IO Int64
+returnMessage copier res = case res of
+    Nothing -> return $ queryResultCode QRNotFound
+    Just v -> do
+        let encoded = Proto.encodeMessage (toProto v)
+        BS.unsafeUseAsCStringLen encoded (\(ptr, len) -> copier (castPtr ptr) (fromIntegral len))
+        return $ queryResultCode QRSuccess
+
+enqueueMessages :: (Proto.Message (Output a), ToProto a) => (Ptr Word8 -> Int64 -> IO Int32) -> [a] -> IO ThreadId
+enqueueMessages callback = enqueueProtoMessages callback . fmap toProto
+
+{- |Spawn a new thread that will invoke the provided callback on the list of
+ encoded messages. If the callback response indicates that the channel to
+ which the callback is enqueueing is full, the thread will wait. The wait time
+ follows exponential backoff strategy to a maximum of 10 seconds.
+-}
+enqueueProtoMessages :: (Proto.Message a) => (Ptr Word8 -> Int64 -> IO Int32) -> [a] -> IO ThreadId
+enqueueProtoMessages callback = forkIO . go 0 . map Proto.encodeMessage
+  where
+    go _ [] = () <$ callback nullPtr maxBound -- close the sender channel.
+    go n msgs@(msg : msgs') =
+        BS.unsafeUseAsCStringLen msg $ \(headPtr, len) -> do
+            res <- callback (castPtr headPtr) (fromIntegral len)
+            if res == 0
+                then go 0 msgs' -- reset wait time to 10ms.
+                else
+                    if res == -1
+                        then do
+                            let delay = (2 ^ n) * 10_000
+                            threadDelay delay
+                            go (min (n + 1) (10 :: Int)) msgs -- maximum delay is 10 seconds
+                        else return () -- the sender channel is now dropped, so we stop.
+
+-- V2 stuff starts here:
```

Comment:
@abizjak The changes start here, above is just lifted (temporarily) from GRPC2.hs of the client.